### PR TITLE
SLING-10469 - detecting dir folder and substructure belonging to nt f…

### PR DESCRIPTION
**Pull request by DominikSuess**
…ile of a configuration and ignore in Handler ot prevent leftover in converted package

### CODE REVIEW

**Comment by bosschaert** [comment]
While I'm not super familiar with this codebase, the change looks reasonable to me.

**Comment by bosschaert** [comment]
@DominikSuess You need to look at the failing CI/CD though...

**Comment by rombert** [comment]
I kicked off another build - the old one seemed transient - and now we have an actual test failure at https://ci-builds.apache.org/job/Sling/job/modules/job/sling-org-apache-sling-feature-cpconverter/job/PR-88/2/testReport/junit/org.apache.sling.feature.cpconverter.handlers/ConfigEntryHandlerTest/?cloudbees-analytics-link=scm-reporting%2Ftests%2Ffailed

**Comment by DominikSuess** [comment]
@bosschaert @rombert - sorry, turns out I turned off the rat test and missed to commit the last little change for the subfolders. This now also passes with rat check enabled for me locally.

**Comment by karlpauls** [comment]
@DominikSuess, can you please rebase the PR and resolve the conflicts? I think we are getting close to a new release and I would like to include this patch.

**Review by anchela — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/AbstractConfigurationEntryHandler.java [hunk 1]
```diff
@@ -33,7 +33,7 @@
     private boolean enforceConfigurationBelowConfigFolder;
 
     public AbstractConfigurationEntryHandler(@NotNull String extension) {
-        super("/jcr_root/(?:apps|libs)/.+/(?<foldername>config|install)(\\.(?<runmode>[^/]+))?/(?<pid>.*)\\." + extension);
+        super("/jcr_root/(?:apps|libs)/.+/(?<foldername>config|install)(\\.(?<runmode>[^/]+))?(.*)/(?<pid>[^\\/]*)\\." + extension + ("(?<dir>.dir(/\\.content\\.xml)?)?$"));
     }
 
     void setEnforceConfgurationBelowConfigFolder(boolean enforceConfigurationBelowConfigFolder) {
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/AbstractConfigurationEntryHandler.java:39]: with the additional if for 'dir' the method gets quite complex. i would suggest to refactor it to reduce overall complexity

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/AbstractConfigurationEntryHandler.java [hunk 2]
```diff
@@ -33,7 +33,7 @@
     private boolean enforceConfigurationBelowConfigFolder;
 
     public AbstractConfigurationEntryHandler(@NotNull String extension) {
-        super("/jcr_root/(?:apps|libs)/.+/(?<foldername>config|install)(\\.(?<runmode>[^/]+))?/(?<pid>.*)\\." + extension);
+        super("/jcr_root/(?:apps|libs)/.+/(?<foldername>config|install)(\\.(?<runmode>[^/]+))?(.*)/(?<pid>[^\\/]*)\\." + extension + ("(?<dir>.dir(/\\.content\\.xml)?)?$"));
```
[the remark this answers is not in the document]
Comment (DominikSuess) [reply src/main/java/org/apache/sling/feature/cpconverter/handlers/AbstractConfigurationEntryHandler.java:36]: in rebased PR I added a corresponding comment - the groups should explain what I'm trying to capture but inline comments now clarify why we watch for it and why we need to ignore those .dir folders

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/AbstractConfigurationEntryHandler.java (hunk 1 above)
Comment (DominikSuess) [reply src/main/java/org/apache/sling/feature/cpconverter/handlers/AbstractConfigurationEntryHandler.java:39]: no objecting a refactoring but  I only added an exit condition for the dir case - This shouldn't be block merging this bugfix in and can hopefully wait until I'm back from PTO

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/AbstractConfigurationEntryHandler.java [hunk 3]
```diff
@@ -48,47 +48,52 @@ public final void handle(@NotNull String path, @NotNull Archive archive, @NotNul
         String runMode;
         // we are pretty sure it matches, here
```
[the remark this answers is not in the document]
Comment (DominikSuess) [reply src/main/java/org/apache/sling/feature/cpconverter/handlers/AbstractConfigurationEntryHandler.java:49]: that's not my additon but the reason is that the  accept method is being called with the same regexp before, so calling it a second time should still match ;)

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/AbstractConfigurationEntryHandler.java [hunk 4]
```diff
@@ -48,47 +48,52 @@ public final void handle(@NotNull String path, @NotNull Archive archive, @NotNul
...
-            String id;
-            int n = pid.indexOf('~');
-            if (n == -1) {
-                n = pid.indexOf('-');
-            }
-            if (n > 0) {
-                factoryPid = pid.substring(0, n);
-                id = factoryPid.concat("~").concat(pid.substring(n + 1));
+            if (matcher.group("dir") != null) {
+                // preventing invalid results as the corresponding configuration will be stripped from the resulting package causing the
```
[the remark this answers is not in the document]
Comment (DominikSuess) [reply src/main/java/org/apache/sling/feature/cpconverter/handlers/AbstractConfigurationEntryHandler.java:52]: completed in the rebased replacement PR

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/AbstractConfigurationEntryHandler.java [hunk 5]
```diff
@@ -48,47 +48,52 @@ public final void handle(@NotNull String path, @NotNull Archive archive, @NotNul
...
+            if (matcher.group("dir") != null) {
+                // preventing invalid results as the corresponding configuration will be stripped from the resulting package causing the 
+                logger.info("{} is only a dir folder next to config - removing.", path);
             } else {
-                id = pid;
-            }
+                String pid = matcher.group("pid");
     
-            logger.info("Processing configuration '{}'.", id);
+                int idx = pid.lastIndexOf('/');
```
[the remark this answers is not in the document]
Comment (DominikSuess) [reply src/main/java/org/apache/sling/feature/cpconverter/handlers/AbstractConfigurationEntryHandler.java:57]: I didn'T change this part but the regexp should only match IF a pid is present

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/AbstractConfigurationEntryHandler.java [hunk 6]
```diff
@@ -48,47 +48,52 @@ public final void handle(@NotNull String path, @NotNull Archive archive, @NotNul
...
+                logger.info("Processing configuration '{}'.", id);
+        
+                Dictionary<String, Object> configurationProperties;
+                try (InputStream input = Objects.requireNonNull(archive.openInputStream(entry))) {
+                    configurationProperties = parseConfiguration(id, input);
+                }
+        
+                if (configurationProperties == null) {
+                    logger.info("{} entry does not contain a valid OSGi configuration, treating it as a regular resource", path);
+                    Objects.requireNonNull(converter.getMainPackageAssembler()).addEntry(path, archive, entry);
```
[the remark this answers is not in the document]
Comment (DominikSuess) [reply src/main/java/org/apache/sling/feature/cpconverter/handlers/AbstractConfigurationEntryHandler.java:83]: took over changed code during rebase

**Comment by DominikSuess** [comment]
Replaced by rebased (and merged) PR https://github.com/apache/sling-org-apache-sling-feature-cpconverter/pull/96

**Pull request closed without merging**